File: JavaScript/recursion/validparentheses.js
/* Implement an algorithm to print all valid (e.g., properly opened and closed) combinations
   of n pairCount of parentheses. : Implement an algorithm to print all valid (e.g., properly opened and closed) combinations
   of n pairCount of parentheses. 
 */

function generateParens(pairCount) {
  function addParen(arr, leftRem, rightRem, chars, index) {
    if (leftRem < 0 || rightRem < leftRem) return;

    if (leftRem === 0 && rightRem === 0) {
      arr.push(chars.join(""));
    } else {
      chars[index] = '(';
      addParen(arr, leftRem - 1, rightRem, chars, index + 1);

      chars[index] = ')';
      addParen(arr, leftRem, rightRem - 1, chars, index + 1);
    }
  }

  let chars = new Array(pairCount * 2);
  let parens = [];
  addParen(parens, pairCount, pairCount, chars, 0)
  return parens;
}

File: Java/algorithms/src/recursion/ValidParentheses.java
package recursion;

import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

/*
 * Problem statement: : Implement an algorithm to print all valid (e.g., properly opened and 
 * closed) combinations of n pairs of parentheses.
 */

public class ValidParentheses {

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		System.out.print("Enter number of valid pairs to generate: ");
		int pairs = scanner.nextInt();
		System.out.println(generateParens(pairs));
		scanner.close();
	}

	private static Set<String> generateParens(int remainingPairs) {
		Set<String> uniqueSet = new HashSet<String>();
		if (remainingPairs == 0) {
			uniqueSet.add("");
		} else {
			Set<String> prevSet = generateParens(remainingPairs - 1);
			for (String str : prevSet) {
				for (int i = 0; i < str.length(); i++) {
					if (str.charAt(i) == '(') {
						String newCombo = insertIn(str, i);
						uniqueSet.add(newCombo);
					}
				}
				uniqueSet.add("()" + str);  //  Add a pair to the beginning of the string 
			}
		}
		return uniqueSet;
	}
	
	// Insert after a left parentheses
	private static String insertIn(String str, int leftParenIdx) {
		String leftSubstr = str.substring(0, leftParenIdx + 1);
		String rightSubstr = str.substring(leftParenIdx + 1);
		return leftSubstr + "()" + rightSubstr;
	}
}
/*
 * (Not very efficient solution)
 */